### PR TITLE
Refactor CLI entry point

### PR DESCRIPTION
**Pull request by bartlomieju**
This PR refactors `main.rs` to use dedicated functions for different `deno <subcommand>`, as [discussed](https://github.com/denoland/deno/pull/2084#discussion_r275699462) in #2084

EDIT: This PR should be blocked by #2113 

EDIT2: As suggested by @kevinkassimo in #2102 and changed `--types` and `--prefetch` to respective subcommands:
```
deno --types -> deno types
deno --prefetch -> deno prefetch
```

### CODE REVIEW

File: cli/main.rs [hunk 1]
```diff
@@ -123,103 +124,177 @@ pub fn print_file_info(worker: &Worker, url: &str) {
...
+      .and_then(move |worker| {
+        print_file_info(&worker, &main_module);
+        worker.then(|result| {
+          js_check(result);
+          Ok(())
+        })
+      }).map_err(|(err, _worker)| print_err_and_exit(err))
+  });
+  tokio_util::run(main_future);
+}
```
Comment (bartlomieju) [inline cli/main.rs]: @ry PTAL at it. It seem that `execute_mod_async` does not evaluate and run file if `prefetch` is set to true, so simple letting worker finish the job seems to work fine. I'm not sure though

File: cli/flags.rs [hunk 2]
```diff
@@ -198,46 +179,7 @@ fn create_cli_app<'a, 'b>() -> App<'a, 'b> {
...
-          .unwrap()
-          .map(String::from)
-          .collect();
-        rest_argv.extend(script_args);
-      }
-    }
-    _ => {}
-  }
-
+pub fn set_flags(matches: ArgMatches) -> Result<DenoFlags, String> {
```
Comment (ry) [inline cli/flags.rs]: Regarding the return value - it doesn't seem like it ever errors? Can this be just
```rust
pub fn set_flags(matches: ArgMatches) -> DenoFlags
```

File: cli/flags.rs [hunk 3]
```diff
@@ -256,138 +198,125 @@ pub fn set_flags(
   }
```
Comment (ry) [inline cli/flags.rs]: `v8_set_flags` is called here, which makes it not a pure parsing function. Can that be moved elsewhere?

File: cli/flags.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/flags.rs]: Yep, will do

File: cli/flags.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/flags.rs]: So based on this [comment](https://github.com/denoland/deno/pull/2113#issuecomment-485135260) I'm gonna add `v8_help` and `v8_flags` fields to `DenoFlags` and call `v8_set_flags` in `main.rs::main()`

Comment (ry) [reply cli/flags.rs]: Yes, sounds good.

File: cli/flags.rs [hunk 4]
```diff
@@ -21,11 +21,6 @@ pub struct DenoFlags {
...
   pub allow_high_precision: bool,
   pub no_prompts: bool,
-  pub types: bool,
-  pub prefetch: bool,
-  pub info: bool,
-  pub fmt: bool,
-  pub eval: bool,
 }
 
 impl<'a> From<ArgMatches<'a>> for DenoFlags {
```
Comment (ry) [inline cli/flags.rs]: Can this `DenoFlags::<ArgMatches>::from` implementation be combined into `set_flags`? I feel like parsing is now being broken up into two parts.

File: cli/flags.rs (hunk 2 above)
Comment (ry) [reply cli/flags.rs]: Can you also rename this function into something like "parse_flags" ?

File: cli/main.rs [hunk 5]
```diff
@@ -123,103 +126,212 @@ pub fn print_file_info(worker: &Worker, url: &str) {
...
+  log::set_logger(&LOGGER).unwrap();
+  let args: Vec<String> = env::args().collect();
+  let cli_app = create_cli_app();
+  let matches = cli_app.get_matches_from(args);
+  let flags = flags::parse_flags(matches.clone());
+
+  if flags.v8_help {
+    // show v8 help and exit
+    // TODO(bartlomieju): this relies on `v8_set_flags` to swap `--v8-options` for "--help"
+    v8_set_flags(vec!["--v8-options".to_string()]);
```
Comment (bartlomieju) [inline cli/main.rs]: Other than this file only `core/examples/http_bench.rs` uses `v8_set_flags`. I guess we can now rid it of `v8_set_flags_preprocess` and remove this obscure swapping logic?

Comment (ry) [reply cli/main.rs]: @bartlomieju Nice clean ups. Getting rid `v8_set_flags_preprocess` seems good. Ping me when I should review again.

Comment (bartlomieju) [reply cli/main.rs]: Thanks @ry. This is now ready for review.

File: cli/flags.rs [hunk 6]
```diff
@@ -197,197 +132,205 @@ fn create_cli_app<'a, 'b>() -> App<'a, 'b> {
...
-  assert_eq!(
-    flags,
-    DenoFlags {
-      log_debug: true,
-      reload: true,
-      ..DenoFlags::default()
-    }
-  );
-}
+  fn flags_from_vec(args: Vec<String>) -> DenoFlags {
```
Comment (ry) [inline cli/flags.rs]: I feel like this function should be the high level public “parse_flags” while the other function which is currently called “parse_flags” should be either combined into this or called something else (and be internal to this module)

File: cli/flags.rs [hunk 7]
```diff
@@ -197,197 +132,205 @@ fn create_cli_app<'a, 'b>() -> App<'a, 'b> {
...
-}
+    let flags =
+      flags_from_vec(svec!["deno", "--v8-flags=--expose-gc,--gc-stats=1"]);
+    assert_eq!(
+      flags,
+      DenoFlags {
+        v8_flags: Some(svec!["--expose-gc", "--gc-stats=1"]),
+        ..DenoFlags::default()
+      }
+    );
```
Comment (ry) [inline cli/flags.rs]: 👍

File: cli/main.rs [hunk 8]
```diff
@@ -123,103 +126,211 @@ pub fn print_file_info(worker: &Worker, url: &str) {
   }
 }
 
-fn main() {
-  #[cfg(windows)]
-  ansi_term::enable_ansi_support().ok(); // For Windows 10
+fn get_worker_and_state(
```
Comment (ry) [inline cli/main.rs]: “get” doesn’t sound right here since the worker and state are being constructed. I’d prefer “create”

File: cli/main.rs [hunk 9]
```diff
@@ -123,103 +126,211 @@ pub fn print_file_info(worker: &Worker, url: &str) {
...
+      let code: &str = info_match.value_of("code").unwrap();
+      argv.extend(vec![code.to_string()]);
+      eval_command(flags, argv);
+    }
+    ("info", Some(info_match)) => {
+      let file: &str = info_match.value_of("file").unwrap();
+      argv.extend(vec![file.to_string()]);
+      prefetch_or_info_command(flags, argv, true);
+    }
+    ("prefetch", Some(info_match)) => {
```
Comment (ry) [inline cli/main.rs]: info_match should be prefetch_match

File: cli/flags.rs (hunk 6 above)
Comment (bartlomieju) [reply cli/flags.rs]: Yeah, I had similar feeling about it, but that means we no longer have `matches` var available in `main()`, that gives nice switch on `matches.subcommand()`.

Comment (bartlomieju) [reply cli/flags.rs]: Alternatively `parse_flags` may return `DenoFlags, ArgMatches` to handle both cases

**Comment by J2P** [comment]
The manual.md file also seems to need to be modified.

**Comment by ry** [comment]
PS we should add a "deno help" command - it feels like that would be expected now.

**Comment by bartlomieju** [comment]
@ry no problem, I can add it in an hour or so. I understand you want to add `deno help` as an alias for `deno --help`, right?

**Pull request merged**